src/sql/src/catalog.rs

```diff
@@ -1328,6 +1330,52 @@ impl<'a, T> ErsatzCatalog<'a, T> {
     }
 }
 
+/// Provides a method of generating a 2-layer catalog on the fly, and then
+/// resolving objects within it.
+pub(crate) struct SubsourceCatalogDouble<'a, T>(pub BTreeMap<String, BTreeMap<String, &'a T>>);
```

Comment:
I think this double version goes a bit against the abstraction (that granted is not well spelled out anywhere yet). The idea is that subsource planning should be handled identically by requiring that each source implementation defines and exports a subsource catalog with a canonical structure. This structure must be in the form of a three layered catalog even if the upstream source has a simpler structure. In that case we can generate a default name for the layer that's missing.

As an example the load generator source, for which no upstream system exists, but it still reports a fully fledged three layer catalog. Can something like this be done for MySQL?

Reply:
After doing that you can also only keep a `SubsourceCatalog<T>` struct and get rid of the trait

Reply:
> I would go for a default name like "mysql" or something like that. 

Hmm, then that would leak `mysql` details into the `impl SubsourceCatalog` implementation of `resolve`..

> The resolve method should already support resolving names based on just the item name or just the item and the schema so long as they are non-ambiguous

Yep it supports that, but we don't want to return a new `UnresolvedItemName` with the extra `mysql` prefix in there, so I think we still need to check if the input only contained 2 `Ident`s and used a default catalog layer, right?

Reply:
Okay got it - thanks for clarifying. We'll have to be careful to resolve things the same way on the source side, so I'll tackle that when I rebase #24203 against this PR. I've implemented this in the commit I'm about to push.